// @storybook/react-webpack5/dist/preset.js
import CJS_COMPAT_NODE_URL_1xanwf2ju77 from 'node:url';
import CJS_COMPAT_NODE_PATH_1xanwf2ju77 from 'node:path';
import CJS_COMPAT_NODE_MODULE_1xanwf2ju77 from "node:module";

var __filename = CJS_COMPAT_NODE_URL_1xanwf2ju77.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_1xanwf2ju77.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_1xanwf2ju77.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------

// src/preset.ts
import { fileURLToPath } from "node:url";
import { WebpackDefinePlugin } from "@storybook/builder-webpack5";
var addons = [
  fileURLToPath(import.meta.resolve("@storybook/preset-react-webpack"))
], core = async (config, options) => {
  let framework = await options.presets.apply("framework");
  return {
    ...config,
    builder: {
      name: fileURLToPath(import.meta.resolve("@storybook/builder-webpack5")),
      options: typeof framework == "string" ? {} : framework.options.builder || {}
    },
    renderer: fileURLToPath(import.meta.resolve("@storybook/react/preset"))
  };
}, webpack = async (config, options) => (config.resolve = config.resolve || {}, config.resolve.alias = {
  ...config.resolve?.alias,
  "@storybook/react": fileURLToPath(import.meta.resolve("@storybook/react"))
}, options.features?.developmentModeForBuild && (config.plugins = [
  // @ts-expect-error Ignore this error, because in the `webpack` preset the user actually hasn't defined a config yet.
  ...config.plugins,
  new WebpackDefinePlugin({
    NODE_ENV: JSON.stringify("development")
  })
]), config);
export {
  addons,
  core,
  webpack
};
